runtime/hydradx/src/lib.rs

```diff
@@ -1186,4 +1181,66 @@ impl_runtime_apis! {
 			Default::default()
 		}
 	}
+	impl sp_api::Core<Block> for Runtime {
+		fn version() -> RuntimeVersion {
+			VERSION
+		}
+
+		fn execute_block(block: Block) {
+			Executive::execute_block(block)
+		}
+
+		fn initialize_block(header: &<Block as BlockT>::Header) -> ExtrinsicInclusionMode {
+			Executive::initialize_block(header)
+		}
+	}
+	impl primitives::runtime_api::FeeEstimationApi<Block, AccountId, AssetId, Balance> for Runtime {
+		fn estimate_fee_payment(weight: Weight, account_id: AccountId) -> (AssetId, Balance) {
+			// Get the currency configured for this account
+			let currency = MultiTransactionPayment::account_currency(&account_id);
+
+			// Convert weight to base fee using native asset calculations
+			let native_fee = TransactionPayment::weight_to_fee(weight);
+
+			// Get the price of the account's currency if applicable
+			let price = MultiTransactionPayment::price(currency).unwrap_or(Price::one());
+
+			// Calculate the final fee in the account's currency
+			let fee_in_currency = price.checked_mul_int(native_fee).unwrap_or(native_fee);
+
+			(currency, fee_in_currency)
+		}
+
+		fn estimate_fee_payment_for_extrinsic(
+			uxt: BoundedVec<u8, MaxExtrinsicSize>,
+			account_id: AccountId,
+		) -> (AssetId, Balance) {
+			use codec::DecodeAll;
+
+			// Decode the extrinsic
+			let uxt = UncheckedExtrinsic::decode_all(&mut &uxt[..])
+				.expect("Invalid extrinsic provided");
```

Comment:
this could cause panics easily.

there is not need to have boundev and and decode stuff. See my other comment to use differetn type for the paramter.